java-api/src/main/kotlin/io/spine/validation/api/MessageValidator.kt

```diff
@@ -0,0 +1,141 @@
+ *
+ * ## Validation of external messages
+ *
+ * The Validation library provides a mechanism that allows validating of
+ * the external messages, **which are used for fields within local messages**.
+ * Implement this interface and annotate the implementing class with
+ * the [@Validator][Validator] annotation, specifying the message type to validate.
+ *
+ * For each field of type [M] within any local message, the library will invoke
+ * the [MessageValidator.validate] method when validating the local message.
+ *
+ * The following Protobuf field types are supported:
+ *
+ * 1. Singular fields of type [M].
+ * 2. Repeated field of type [M].
+ * 3. Map field with values of type [M].
+ *
+ * An example of the validator declaration for the `Earphones` message:
+ *
+ * ```kotlin
+ * @Validator(Earphones::class)
+ * public class EarphonesValidator : MessageValidator<Earphones> {
+ *     public override fun validate(message: Earphones): List<DetectedViolation> {
+ *         return emptyList() // Always valid.
+ *     }
+ * }
+ * ```
+ *
+ * Please note that standalone instances of [M] and fields of [M] type that occur in
+ * other external messages **will not be validated**.
+ *
+ * Consider the following example:
+ *
+ * ```proto
+ * // Brings the `Earphones` message from dependencies.
+ * // Suppose we don't control the generated code of declarations from this file.
+ * import "earphones.proto";
+ *
+ * // A locally-declared message.
+ * message WorkingSetup {
+ *
+ *     // The field of external message type.
+ *     Earphones earphones = 1;
+ * }
+ * ```
+ *
+ * Supposing that the Validation library applied to the module where both `WorkingSetup`
+ * and `EarphonesValidator` classes are declared, then the generated code of `WorkingSetup`
+ * will apple the validator to each instance passed to the `WorkingSetup.earphones` field.
+ *
+ * ## Implementation
+ *
+ * The message validator does not have restrictions upon how exactly the message
+ * must be validated. It can validate a particular field, several fields,
+ * the whole message instance (for example, checking the field relations),
+ * and perform a deep validation.
+ *
+ * It is a responsibility of the validator to provide the correct instances
+ * of [DetectedViolation]. Before reporting to the user, the library converts
+ * [DetectedViolation] to a [ConstraintViolation][io.spine.validate.ConstraintViolation].
+ * Returning of an empty list of violations means that the given message is valid.
+ *
+ * Please keep in mind that for each invocation a new instance of [MessageValidator]
+ * is created. Every implementation of [MessageValidator] must have a public,
+ * no-args constructor.
+ *
+ * An implementation of [MessageValidator] will be rejected by the library
+ * in the following cases:
+ *
+ * 1) It is used to validate a local message. Only external messages are allowed
+ *    to have a validator.
+ * 2) There already exists a validator for the specified message type. Having several
```

Comment:
Let's also describe the scope in which these rules apply. E.g. one may think it's scoped by a Gradle module, while it is certainly not.

Please have it somewhere above, as it is an important point.